pkg/rkeworker/execute_windows.go

```diff
@@ -1,106 +1,69 @@
 package rkeworker
 
 import (
-	"bytes"
 	"context"
-	"encoding/base64"
-	"errors"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"os"
 	"os/exec"
 	"path/filepath"
 	"strings"
-	"syscall"
 
-	"github.com/rancher/norman/types"
-	"github.com/rancher/rancher/pkg/rkecerts"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/sync/errgroup"
 )
 
-var (
-	rancherDir = `C:\etc\rancher`
+var ErrHyperKubePSScriptAgentRetry = errors.New("PowerShell process is going to retry")
 
-	ErrHyperKubePSScriptCrash      = errors.New("crashed to execute PowerShell process")
-	ErrHyperKubePSScriptAgentRetry = errors.New("PowerShell process is going to retry")
-)
-
-func ExecutePlan(ctx context.Context, nodeConfig *NodeConfig, writeCertOnly bool) error {
-	if nodeConfig.Certs != "" {
-		bundle, err := rkecerts.Unmarshal(nodeConfig.Certs)
-		if err != nil {
-			return err
-		}
-
-		if err := bundle.Explode(); err != nil {
-			return err
-		}
-	}
-
-	f := fileWriter{}
-	for _, file := range nodeConfig.Files {
-		f.write(file.Name, file.Contents)
-	}
-	if writeCertOnly {
-		return nil
-	}
-
-	// pre run docker
-	for name, process := range nodeConfig.Processes {
-		if strings.HasPrefix(name, "pre-run-docker") {
-			if err := runProcess(ctx, process.Name, process, true, false); err != nil {
-				return err
+func init() {
+	doExecutePlan = func(ctx context.Context, nodeConfig *NodeConfig, _ bool) error {
+		// run as docker
+		for name, process := range nodeConfig.Processes {
+			if strings.HasPrefix(name, "run-container-") {
+				if err := runProcess(ctx, process.Name, process, true, false); err != nil {
+					return err
+				}
 			}
 		}
-	}
 
-	// post run docker
-	for name, process := range nodeConfig.Processes {
-		if strings.HasPrefix(name, "post-run-docker") {
-			if err := runProcess(ctx, process.Name, process, true, false); err != nil {
-				return err
-			}
-		}
-	}
+		// run as powershell
+		for name, process := range nodeConfig.Processes {
+			if strings.HasPrefix(name, "run-script-") {
+				scriptPath, err := getPowerShellScript(process.Name)
+				if err != nil {
+					return err
+				}
 
-	// run hyperkube.ps1
-	if hyperkubeProcess, ok := nodeConfig.Processes["hyperkube"]; !ok {
-		return errors.New("can't execute hyperkube.ps1 without any commands")
-	} else {
-		scriptPath := rancherDir + `\hyperkube.ps1`
-		if _, err := os.Stat(scriptPath); err != nil {
-			return fmt.Errorf(`can't find %s'`, scriptPath)
+				if err = executePowerShell(ctx, scriptPath, process.Args...); err != nil {
+					return err
+				}
+			}
 		}
 
-		return executePowerShell(ctx, scriptPath, hyperkubeProcess.Args...)
+		return nil
 	}
 }
 
 func Stop(ctx context.Context) error {
-	scriptPath := rancherDir + `\stop.ps1`
-	if _, err := os.Stat(scriptPath); err != nil {
-		return fmt.Errorf(`can't find %s while stopping`, scriptPath)
+	scriptPath, err := getPowerShellScript("stop")
```

Comment:
You are following the same pattern.
```
1. getPowerShellScript
2. executePowerShell
```
Maybe we can combine both function together.

Reply:
updated